glide-core/src/client/value_conversion.rs

```diff
@@ -673,6 +674,289 @@ pub(crate) fn convert_to_expected_type(
+                          2) (integer) 1719710707260
+                          3) (integer) 1
+                       2) 1) "1719710688676-0"
+                          2) (integer) 1719710718373
+                          3) (integer) 1
+        
+        Another RESP3 example on an empty stream
+
+        1# "length" => (integer) 0
+        2# "radix-tree-keys" => (integer) 0
+        3# "radix-tree-nodes" => (integer) 1
+        4# "last-generated-id" => "0-1"
+        5# "max-deleted-entry-id" => "0-1"
+        6# "entries-added" => (integer) 1
+        7# "recorded-first-entry-id" => "0-0"
+        8# "entries" => (empty array)
+        9# "groups" => (empty array)
+
+        Without `FULL` keyword, command returns "first-entry" and "last-entry" instead of "entries" in the same format.
+
+        RESP2:
+
+        1) "length"
+        2) (integer) 2
+        3) "radix-tree-keys"
+        4) (integer) 1
+        5) "radix-tree-nodes"
+        6) (integer) 2
+        7) "last-generated-id"
+        8) "1719710688676-0"
+        9) "max-deleted-entry-id"
+        10) "0-0"
+        11) "entries-added"
+        12) (integer) 2
+        13) "recorded-first-entry-id"
+        14) "1719710679916-0"
+        15) "groups"
+        16) (integer) 1
+        17) "first-entry"
+        18) 1) "1719710679916-0"
+            2) 1) "foo"
+               2) "bar"
+               3) "foo"
+               4) "bar2"
+               5) "some"
+               6) "value"
+        19) "last-entry"
+        20) 1) "1719710688676-0"
+            2) 1) "foo"
+               2) "bar2"
+
+        RESP3 on an empty stream:
+
+        1# "length" => (integer) 0
+        2# "radix-tree-keys" => (integer) 0
+        3# "radix-tree-nodes" => (integer) 1
+        4# "last-generated-id" => "0-1"
+        5# "max-deleted-entry-id" => "0-1"
+        6# "entries-added" => (integer) 1
+        7# "recorded-first-entry-id" => "0-0"
+        8# "groups" => (integer) 0
+        9# "first-entry" => (nil)
+        10# "last-entry" => (nil)
+
+        So we convert:
+        - Arrays to maps, accroding to RESP2 -> RESP3 conversion done by the server:
+          - Top level array - unflat an `obj[]` (`[k, v, k, v, ...]`) to a `Map<str, obj>`
+          - "groups" value - `obj[][]` -> `Map<str, obj>[]` - unflat every nested array the same manner as above
+          - "consumers" value - `obj[][]` -> `Map<str, obj>[]` - same as above
+        - Additionally we convert some map's values:
+          - "entries", "first-entry" and "last-entry" value - to a `Map<str, str[][]>` (similar to `XREAD`)
+              no nested maps due to duplicating keys - see example
+        Using `XInfoStreamFullReturnType` recursively for maps' value type.
+        */
+        ExpectedReturnType::XInfoStreamFullReturnType => match value {
+            Value::Map(_) => Ok(value),  // Response is already in RESP3 format - no conversion needed
+            Value::Array(mut array) => {
```

Comment:
Could probably refactor some of this code, since it's repetitive in some areas. We're essentially using the array like an association list, so we probably could write a (zero-cost) abstraction around this behaviour to make things a bit easier to reason about. Not a blocker, but something to think about.

Reply:
Yes, not sure if we will get around to it in this PR but we'll see. I agree that we can probably do some refactoring here to improve things